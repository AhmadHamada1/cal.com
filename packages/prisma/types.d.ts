declare global {
  namespace PrismaJson {
    type EventTypeMetaData = any;
  }
}

declare module "@prisma/client" {
  namespace Prisma {
    interface EventTypeCreateInput {
      customReplyToEmail?: string | null;
      hideOrganizerEmail?: boolean;
      includeNoShowInRRCalculation?: boolean;
    }

    interface EventTypeUpdateInput {
      customReplyToEmail?: string | null;
      hideOrganizerEmail?: boolean;
      includeNoShowInRRCalculation?: boolean;
    }

    interface EventTypeSelect {
      customReplyToEmail?: boolean;
      hideOrganizerEmail?: boolean;
      includeNoShowInRRCalculation?: boolean;
    }

    interface MembershipCreateInput {
      createdAt?: Date | string | null;
    }

    interface MembershipSelect {
      createdAt?: boolean;
    }

    interface OrganizationSettingsUpdateInput {
      disablePhoneOnlySMSNotifications?: boolean;
    }

    interface OrganizationSettingsSelect {
      disablePhoneOnlySMSNotifications?: boolean;
    }

    interface CredentialPayload {
      delegationCredentialId?: string | null;
    }

    interface CredentialSelect {
      delegationCredentialId?: boolean;
    }

    interface CalendarCacheWhereInput {
      userId?: number | Prisma.IntFilter<Prisma.CalendarCacheScalarWhereInput>;
    }

    interface CalendarCacheUpdateInput {
      userId?: number;
    }

    interface CalendarCacheCreateInput {
      userId?: number;
    }
  }
}
