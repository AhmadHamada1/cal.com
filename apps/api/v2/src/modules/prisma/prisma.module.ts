import { Module } from "@nestjs/common";
import { PrismaReadService } from "@/modules/prisma/prisma-read.service";
import { PrismaWriteService } from "@/modules/prisma/prisma-write.service";

@Module({
  providers: [PrismaReadService, PrismaWriteService],
  exports: [PrismaReadService, PrismaWriteService],
})
export class PrismaModule {}
