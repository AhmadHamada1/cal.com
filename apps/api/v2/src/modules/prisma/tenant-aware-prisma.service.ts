import { Injectable, Scope } from "@nestjs/common";
import { Inject } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { REQUEST } from "@nestjs/core";
import { PrismaClient } from "@prisma/client";

import { getTenantFromHost } from "@calcom/prisma/store/tenants";

@Injectable({ scope: Scope.REQUEST })
export class TenantAwarePrismaService {
  public prisma: PrismaClient;

  constructor(
    @Inject(REQUEST) private readonly request: Request,
    private readonly configService: ConfigService
  ) {
    const host = this.request.headers["host"] || "";
    const tenant = getTenantFromHost(host);

    let dbUrl = this.configService.get("db.url", { infer: true });
    if (tenant === "eu") {
      dbUrl = this.configService.get("db.euUrl", { infer: true });
    }

    this.prisma = new PrismaClient({
      datasources: {
        db: {
          url: dbUrl,
        },
      },
    });
  }
}
