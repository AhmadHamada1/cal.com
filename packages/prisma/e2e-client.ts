import { getPrismaClient } from "./store/prismaStore";

export const prismaClientForTest = getPrismaClient({
  datasources: { db: { url: process.env.DATABASE_URL } },
});
