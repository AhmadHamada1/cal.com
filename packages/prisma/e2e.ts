import { getPrismaClient } from "./store/prismaStore";

let _prisma: ReturnType<typeof getPrismaClient>;
export const getPrisma = () => {
  if (!_prisma) {
    _prisma = getPrismaClient();
  }
  return _prisma;
};
// Only export one instance

export const prisma = getPrisma();
