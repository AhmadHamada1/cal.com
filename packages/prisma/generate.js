const { execSync } = require("child_process");

try {
  execSync("npx prisma generate", { stdio: "inherit" });
  execSync("npx prisma format", { stdio: "inherit" });
} catch (e) {
  console.error("⚠️ Prisma generate failed in postinstall:", e?.message || e);
  // Optional: allow CI to continue, or throw to fail hard
  process.exit(0);
}
