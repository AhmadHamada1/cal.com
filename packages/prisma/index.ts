import type { Prisma } from "@prisma/client";

import type { PrismaClientWithExtensions } from "./store/prismaStore";
import { getPrisma, getTenantAwarePrisma } from "./store/prismaStore";
import { Tenant } from "./store/tenants";

const prismaOptions: Prisma.PrismaClientOptions = {};

const loggerLevel = parseInt(process.env.NEXT_PUBLIC_LOGGER_LEVEL ?? "", 10);

if (!isNaN(loggerLevel)) {
  switch (loggerLevel) {
    case 5:
    case 6:
      prismaOptions.log = ["error"];
      break;
    case 4:
      prismaOptions.log = ["warn", "error"];
      break;
    case 3:
      prismaOptions.log = ["info", "error", "warn"];
      break;
    default:
      // For values 0, 1, 2 (or anything else below 3)
      prismaOptions.log = ["query", "info", "error", "warn"];
      break;
  }
}

// If any changed on middleware server restart is required
// TODO: Migrate it to $extends

export const prisma = new Proxy({} as PrismaClientWithExtensions, {
  get(target, prop) {
    if (process.env.NODE_ENV === "test") {
      const defaultPrisma = getPrisma(Tenant.US, prismaOptions);
      return Reflect.get(defaultPrisma, prop);
    }

    try {
      const tenantPrisma = getTenantAwarePrisma(prismaOptions);
      return Reflect.get(tenantPrisma, prop);
    } catch (error) {
      throw new Error(
        "Prisma was called outside of runWithTenants. Please wrap your code with runWithTenants or use a tenant-aware approach."
      );
    }
  },
});

// This prisma instance is meant to be used only for READ operations.
// If self hosting, feel free to leave INSIGHTS_DATABASE_URL as empty and `readonlyPrisma` will default to `prisma`.
export const readonlyPrisma = process.env.INSIGHTS_DATABASE_URL
  ? getPrisma(Tenant.INSIGHTS, prismaOptions)
  : prisma;

export type PrismaClient = PrismaClientWithExtensions;

type OmitPrismaClient = Omit<
  PrismaClient,
  "$connect" | "$disconnect" | "$on" | "$transaction" | "$use" | "$extends"
>;

// we cant pass tx to functions as types miss match since we have a custom prisma client https://github.com/prisma/prisma/discussions/20924#discussioncomment-10077649
export type PrismaTransaction = OmitPrismaClient;

export default prisma;

export * from "./selects";
