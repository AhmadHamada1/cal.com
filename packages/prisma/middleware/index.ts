export { default as bookingReferenceMiddleware } from "./bookingReference";
export { default as slowQueryDetectionMiddleware } from "./slowQueryDetection";
