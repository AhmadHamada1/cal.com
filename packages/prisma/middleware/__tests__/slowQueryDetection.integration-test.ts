import { captureException } from "@sentry/nextjs";
import { beforeEach, describe, expect, it, vi, afterEach } from "vitest";
import type { Mock } from "vitest";

import prisma from "@calcom/prisma";

import slowQueryDetectionMiddleware, { reportSlowQuery, initializeClientState } from "../slowQueryDetection";

vi.mock("@sentry/nextjs", () => ({
  captureException: vi.fn(),
}));

vi.mock("@calcom/lib/logger", () => ({
  default: {
    warn: vi.fn(),
  },
}));

const originalUse = prisma.$use;
const originalOn = prisma.$on;
const originalTransaction = prisma.$transaction;

describe("Slow Query Detection Middleware - Integration Tests", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    (captureException as unknown as Mock).mockClear();

    process.env.SLOW_QUERY_THRESHOLD_MS = "500";
    process.env.NODE_ENV = "test";

    slowQueryDetectionMiddleware(prisma);

    initializeClientState(prisma);
  });

  afterEach(() => {
    delete process.env.SLOW_QUERY_THRESHOLD_MS;
    delete process.env.NODE_ENV;
  });

  it("should not report fast queries to Sentry", async () => {
    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const startTime = Date.now();
    await prisma.user.findFirst({
      where: { id: { not: null } },
      take: 1,
    });
    const endTime = Date.now();
    const duration = endTime - startTime;

    const params = {
      model: "User",
      action: "findFirst",
      args: { where: { id: { not: null } }, take: 1 },
    };

    const reported = reportSlowQuery(prisma, params, 400, startTime);

    expect(reported).toBe(false);
    expect(captureException).not.toHaveBeenCalled();
  });

  it("should report slow queries to Sentry with SQL details", async () => {
    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const startTime = Date.now();
    await prisma.user.findFirst({
      where: { id: { not: null } },
      include: {
        accounts: true,
        bookings: true,
        credentials: true,
        teams: true,
      },
      take: 1,
    });

    const params = {
      model: "User",
      action: "findFirst",
      args: {
        where: { id: { not: null } },
        include: {
          accounts: true,
          bookings: true,
          credentials: true,
          teams: true,
        },
        take: 1,
      },
    };

    const reported = reportSlowQuery(prisma, params, 700, startTime, {
      forceReport: true, // Force report regardless of actual duration
    });

    expect(reported).toBe(true);
    expect(captureException).toHaveBeenCalledTimes(1);
    expect(captureException).toHaveBeenCalledWith(
      expect.objectContaining({
        message: expect.stringContaining("Slow Prisma Query"),
      }),
      expect.objectContaining({
        extra: expect.objectContaining({
          query: expect.objectContaining({
            action: "findFirst",
            duration: expect.any(Number),
          }),
        }),
        tags: expect.objectContaining({
          type: "slow_query",
        }),
      })
    );
  });

  it("should respect rate limiting and not flood Sentry with reports", async () => {
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const timestamp = Date.now();
    const now = timestamp;

    await prisma.user.findFirst({
      where: { id: { not: null } },
      include: {
        accounts: true,
        bookings: true,
      },
      take: 1,
    });

    const params = {
      model: "User",
      action: "findFirst",
      args: {
        where: { id: { not: null } },
        include: {
          accounts: true,
          bookings: true,
        },
        take: 1,
      },
    };

    const firstReported = reportSlowQuery(prisma, params, 700, timestamp, {
      forceReport: true,
      currentTime: now,
    });

    expect(firstReported).toBe(true);
    expect(captureException).toHaveBeenCalledTimes(1);

    (captureException as unknown as Mock).mockClear();

    const secondReported = reportSlowQuery(prisma, params, 700, timestamp + 1000, {
      currentTime: now + 1000, // Only 1 second later, should be rate limited
    });

    expect(secondReported).toBe(false);
    expect(captureException).not.toHaveBeenCalled();
  });

  it("should use the configured threshold from environment variable", async () => {
    vi.clearAllMocks();
    (captureException as unknown as Mock).mockClear();

    const lowerThreshold = 200;

    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const timestamp = Date.now();

    await prisma.user.findFirst({
      where: { id: { not: null } },
      include: { accounts: true },
      take: 1,
    });

    const params = {
      model: "User",
      action: "findFirst",
      args: {
        where: { id: { not: null } },
        include: { accounts: true },
        take: 1,
      },
    };

    const reported = reportSlowQuery(prisma, params, 350, timestamp, {
      forceReport: true,
      overrideThreshold: lowerThreshold,
    });

    expect(reported).toBe(true);
    expect(captureException).toHaveBeenCalledTimes(1);
  });

  it("should handle complex queries with joins", async () => {
    vi.clearAllMocks();
    (captureException as unknown as Mock).mockClear();

    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const timestamp = Date.now();

    await prisma.booking.findFirst({
      where: { userId: { not: null } },
      include: {
        user: {
          include: {
            teams: true,
          },
        },
        attendees: true,
        references: true,
        payment: true,
      },
      take: 1,
    });

    const params = {
      model: "Booking",
      action: "findFirst",
      args: {
        where: { userId: { not: null } },
        include: {
          user: {
            include: {
              teams: true,
            },
          },
          attendees: true,
          references: true,
          payment: true,
        },
        take: 1,
      },
    };

    const reported = reportSlowQuery(prisma, params, 700, timestamp, {
      forceReport: true,
    });

    expect(reported).toBe(true);
    expect(captureException).toHaveBeenCalledTimes(1);
    expect(captureException).toHaveBeenCalledWith(
      expect.objectContaining({
        message: expect.stringContaining("Slow Prisma Query"),
      }),
      expect.objectContaining({
        extra: expect.objectContaining({
          query: expect.objectContaining({
            model: "Booking",
          }),
        }),
      })
    );
  });

  it("should handle middleware execution correctly", async () => {
    vi.clearAllMocks();
    (captureException as unknown as Mock).mockClear();

    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;

    const timestamp = Date.now();

    const params = {
      model: "User",
      action: "findUnique",
      args: { where: { id: 1 } },
    };

    const middlewareFn = async (params: any, next: any) => {
      const result = await next(params);

      reportSlowQuery(prisma, params, 700, timestamp, {
        forceReport: true,
      });

      return result;
    };

    const next = vi.fn().mockResolvedValue({ id: 1, name: "Test User" });
    await middlewareFn(params, next);

    expect(next).toHaveBeenCalledWith(params);
    expect(captureException).toHaveBeenCalledTimes(1);
  });

  it("should capture and snapshot raw SQL queries", async () => {
    vi.clearAllMocks();
    (captureException as unknown as Mock).mockClear();

    // Reset the last report time to ensure we're not rate limited
    const state = initializeClientState(prisma);
    state.lastReportTime = 0;
    state.queryMap.clear();

    const mockSql = 'SELECT * FROM "User" WHERE "id" IS NOT NULL LIMIT 1';
    const timestamp = Date.now();

    state.queryMap.set("test-query", {
      sql: mockSql,
      timestamp: timestamp,
    });

    const params = {
      model: "User",
      action: "findFirst",
      args: { where: { id: { not: null } }, take: 1 },
    };

    const reported = reportSlowQuery(prisma, params, 700, timestamp, {
      forceReport: true,
    });

    expect(reported).toBe(true);
    expect(captureException).toHaveBeenCalledTimes(1);

    const captureExceptionCall = (captureException as unknown as Mock).mock.calls[0];
    const capturedQuery = captureExceptionCall[1].extra.query;

    expect(capturedQuery.sql).toBeDefined();
    expect(capturedQuery.sql).not.toBe("SQL not captured");

    expect(capturedQuery.sql).toMatchSnapshot();
  });
});
