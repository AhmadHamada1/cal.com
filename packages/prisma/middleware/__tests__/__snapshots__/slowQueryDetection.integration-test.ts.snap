// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should capture and snapshot raw SQL queries 1`] = `"SELECT * FROM \\"User\\" WHERE \\"id\\" IS NOT NULL LIMIT 1"`;
